File: run_600.R

outputfile="results_600.RData"
outputfilecsv="results_600.csv"

for(id in 1:100) {

 jobid <- pcatsAPIclientR::dynamicGP(datafile=paste0("data/600/sim",id,".csv"),
   stg1.outcome="Y1",
   stg1.treatment="A1",
   stg1.x.explanatory="X,Z",
   stg1.x.confounding="X,Z",
   stg1.outcome.type="Continuous",
   stg1.tr.hte="Z",
   stg1.tr.type = "Discrete",
   stg2.outcome="Y2",
   stg2.treatment="A2",
   stg2.x.explanatory="X,Y1,Z",
   stg2.x.confounding="X,Y1,Z",
   stg2.outcome.type="Continuous",
   stg2.tr2.hte="Z",
   stg2.tr.type = "Discrete",
   burn.num=2000,
   mcmc.num=2000,
   x.categorical='Z',
   method="GP",
   seed=1,
   use.cache=T,
   reuse.cached.jobid=T)

   print(jobid)

   status <- pcatsAPIclientR::wait_for_result(jobid)

   if (status$status != "Done") {
      stop(paste("Error for id", id, "status:", status$status))
   }

   result<-pcatsAPIclientR::results(jobid)

   data <- rbind(result$dynamicGP$stage1$ate, result$dynamicGP$stage2$ate)
   data <- cbind(data, var=c("Y1Y0", "Y01Y00", "Y10Y00", "Y11Y00", "Y10Y01", "Y11Y01", "Y11Y10"))
   outdata <-data.frame("ID"=id)
   for(i in 1:7) {
      outdata[,paste0(data[i,]$var,".Estimation")]<-data[i,]$Estimation * -1
      outdata[,paste0(data[i,]$var,".SD")]<-data[i,]$SD
      outdata[,paste0(data[i,]$var,".LB")]<-data[i,]$LB * -1
      outdata[,paste0(data[i,]$var,".UB")]<-data[i,]$UB * -1
   }
   if (file.exists(outputfile)) load(outputfile)
   else results <- NULL

   if (sum(results$ID==id)) {
      results[which(results$ID==id),] <- outdata
   } else {
      results <- rbind(results, outdata)
   }

   save(results,file=outputfile)
   write.csv(results, file=outputfilecsv, row.names=FALSE)
}

File: run_600.py
import csv
import json
import os.path
from os import path
import pcats_api_client as pcats_api

outputfilecsv="pyresults_600.csv"

for id in range(1,10):
   jobid=pcats_api.dynamicgp(datafile="data/600/sim{}.csv".format(id),
      stg1_outcome="Y1",
      stg1_treatment="A1",
      stg1_x_explanatory="X,Z",
      stg1_x_confounding="X,Z",
      stg1_outcome_type="Continuous",
      stg1_tr_hte="Z",
      stg1_tr_type = "Discrete",
      stg2_outcome="Y2",
      stg2_treatment="A2",
      stg2_x_explanatory="X,Y1,Z",
      stg2_x_confounding="X,Y1,Z",
      stg2_outcome_type="Continuous",
      stg2_tr2_hte="Z",
      stg2_tr_type="Discrete",
      burn_num=2000,
      mcmc_num=2000,
      x_categorical='Z',
      method="GP",
      seed=1,
      use_cache=1,
      reuse_cached_jobid=1)

   print("JobID: {}".format(jobid))

   status = pcats_api.wait_for_result(jobid)

   if status!="Done":
       print("Error")
       continue

   result_raw = pcats_api.results(jobid)

   result = json.loads(result_raw)

   data = result['dynamicGP']['stage2']['ate']+result['dynamicGP']['stage1']['ate']

   outdata = dict()
   outdata['Seed'] = id
   outdata['Method'] = "GP (1.0)"
   var=["Y01Y00", "Y10Y00", "Y11Y00", "Y10Y01", "Y11Y01", "Y11Y10", "Y1Y0"]
   for i in range(0,7):
      outdata[var[i]]=data[i]['Estimation'] * -1
      outdata["SD."+var[i]]=data[i]['SD']
      outdata["Lb."+var[i]]=data[i]['UB'] * -1
      outdata["Ub."+var[i]]=data[i]['LB'] * -1

   outdata['wall_time'] = result['timing'][2]
   outdata['cpu_time'] = result['timing'][0]+result['timing'][1]

   infile = list()

   if path.exists(outputfilecsv):
     with open(outputfilecsv, 'r') as csvfile:
       reader = csv.DictReader(csvfile, list(outdata.keys()))
       next(reader, None)  # skip the headers
       for row in reader:
          infile.append(row)

   if len(infile)>=id: 
      infile[id-1] = outdata
   else:
      infile.append(outdata)

   with open(outputfilecsv, 'w') as csvfile:
     writer = csv.DictWriter(csvfile, fieldnames=list(outdata.keys()))
     writer.writeheader()
     for row in infile:
        writer.writerow(row)
